Add frameInfo and sourceFrameInfo arguments to _WKInputDelegate.willSubmitFormValues

arenevier opened the pull request:
#### 244e56e9d5387f5bf3f1a37e276bb1ec68fa5863
<pre>
Add frameInfo and sourceFrameInfo arguments to _WKInputDelegate.willSubmitFormValues
<a href="https://rdar.apple.com/147531053">rdar://147531053</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296504">https://bugs.webkit.org/show_bug.cgi?id=296504</a>

Reviewed by Alex Christensen.

We need to get informations about the frame in the callback. So we pass
the FrameInfoDatas WebPageProxy, then to API::FormClient. And we finally
expose them as WKFrameInfo* to _WKInputDelegate.

* Source/WebKit/UIProcess/API/APIFormClient.h:
(API::FormClient::willSubmitForm):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setInputDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/API/glib/WebKitFormClient.cpp:
* Source/WebKit/UIProcess/WebFormClient.cpp:
(WebKit::WebFormClient::willSubmitForm):
* Source/WebKit/UIProcess/WebFormClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willSubmitForm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchWillSubmitForm):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm:
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[InputDelegateLegacyAPI _webView:willSubmitFormValues:userObject:submissionHandler:]):
(TEST(WebKit, FormSubmission)):
(TEST(WebKit, FormSubmissionLegacyAPI)):
(-[InputDelegate _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/297883@main">https://commits.webkit.org/297883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15949f5686de21590d7fd1164d9ebdb97cd7360f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113242 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23455 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119450 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63903 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115204 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41540 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86189 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116189 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26863 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101867 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66511 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26134 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19995 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63204 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96244 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20070 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122667 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40320 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30079 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95038 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40705 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98082 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94780 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24185 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39926 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17742 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36456 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40206 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45705 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39847 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43180 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41584 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
```diff
@@ -46,6 +47,7 @@ typedef NS_ENUM(NSInteger, _WKFocusStartsInputSessionPolicy) {
 
 - (void)_webView:(WKWebView *)webView didStartInputSession:(id <_WKFormInputSession>)inputSession;
 - (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values userObject:(NSObject <NSSecureCoding> *)userObject submissionHandler:(void (^)(void))submissionHandler;
+- (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values userObject:(NSObject <NSSecureCoding> *)userObject frameInfo:(WKFrameInfo *)frameInfo sourceFrameInfo:(WKFrameInfo *)sourceFrameInfo submissionHandler:(void (^)(void))submissionHandler WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA));
```
We've recently started adding visionos(WK_XROS_TBA) too.
Also, for no particular reason I feel like having the user object after the frame infos would be a little better.  It's typically at the end or right before the completion handler.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -8489,22 +8489,22 @@ void WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation(WebCore::Navig
 
-    RefPtr sourceFrame = WebFrameProxy::webFrame(sourceFrameID);
+    RefPtr sourceFrame = WebFrameProxy::webFrame(sourceFrameInfoData.frameID);
     MESSAGE_CHECK_COMPLETION_BASE(sourceFrame, connection, completionHandler());
```
Let's change this to an early return to match if !frame above.  This seems like a race condition that can be hit with site isolation, and if it is hit we should do nothing rather than terminate the process.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
```diff
@@ -61,8 +61,9 @@ @interface InputDelegateForFormSubmission : NSObject <_WKInputDelegate>
-- (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values userObject:(NSObject <NSSecureCoding> *)userObject submissionHandler:(void (^)(void))submissionHandler
+- (void)_webView:(WKWebView *)webView willSubmitFormValues:(NSDictionary *)values userObject:(NSObject <NSSecureCoding> *)userObject frameInfo:(WKFrameInfo *)frameInfo sourceFrameInfo:(WKFrameInfo *)sourceFrameInfo submissionHandler:(void (^)(void))submissionHandler
 {
+    NSLog(@"abc frameInfo: %@", frameInfo);
```
Remove

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
```diff
@@ -97,18 +100,55 @@ - (void)setShouldStartInputSessionHandler:(BOOL(^)(id <_WKFocusedElementInfo>))h
+
+@end
+
+static NSString* documentWithFormHTML = @"<body onload='document.getElementById(\"formID\").submit()'><form id='formID' method='post' action='test:///formtarget'>"
```
For ObjC types, we want the pointer next to the identifier.

```suggestion
static NSString *documentWithFormHTML = @"<body onload='document.getElementById(\"formID\").submit()'><form id='formID' method='post' action='test:///formtarget'>"
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
```diff
@@ -97,18 +100,55 @@ - (void)setShouldStartInputSessionHandler:(BOOL(^)(id <_WKFocusedElementInfo>))h
+
+TEST(WebKit, FormSubmissionLegacyAPI)
+{
+    auto delegate = adoptNS([[InputDelegateLegacyAPI alloc] init]);
```
RetainPtr to go with our SaferCPP guidelines
```suggestion
    RetainPtr delegate = adoptNS([[InputDelegateLegacyAPI alloc] init]);
```

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
```diff
@@ -97,18 +100,55 @@ - (void)setShouldStartInputSessionHandler:(BOOL(^)(id <_WKFocusedElementInfo>))h
+    auto delegate = adoptNS([[InputDelegateLegacyAPI alloc] init]);
+    WKWebViewConfiguration *configuration = [WKWebViewConfiguration _test_configurationWithTestPlugInClassName:@"BundleFormDelegatePlugInWithUserInfo"];
+    [configuration setURLSchemeHandler:delegate.get() forURLScheme:@"test"];
+    auto webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration]);
```
```suggestion
    RetainPtr webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration]);
```

aprotyas reviewed (commented):
Just a couple of minor style comments

arenevier committed:
Add frameInfo and sourceFrameInfo arguments to _WKInputDelegate.willSubmitFormValues
rdar://147531053
https://bugs.webkit.org/show_bug.cgi?id=296504

Reviewed by Alex Christensen.

We need to get informations about the frame in the callback. So we pass
the FrameInfoDatas WebPageProxy, then to API::FormClient. And we finally
expose them as WKFrameInfo* to _WKInputDelegate.

* Source/WebKit/UIProcess/API/APIFormClient.h:
(API::FormClient::willSubmitForm):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setInputDelegate:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKInputDelegate.h:
* Source/WebKit/UIProcess/API/glib/WebKitFormClient.cpp:
* Source/WebKit/UIProcess/WebFormClient.cpp:
(WebKit::WebFormClient::willSubmitForm):
* Source/WebKit/UIProcess/WebFormClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::willSubmitForm):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchWillSubmitForm):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/DuplicateCompletionHandlerCalls.mm:
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[DuplicateCompletionHandlerCallsDelegate _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UserContentController.mm:
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[InputDelegateForFormSubmission _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/_WKInputDelegate.mm:
(-[InputDelegate _webView:willSubmitFormValues:frameInfo:sourceFrameInfo:userObject:submissionHandler:]):
(-[InputDelegateLegacyAPI _webView:willSubmitFormValues:userObject:submissionHandler:]):
(TEST(WebKit, FormSubmission)):
(TEST(WebKit, FormSubmissionLegacyAPI)):
(-[InputDelegate _webView:willSubmitFormValues:userObject:submissionHandler:]): Deleted.

Canonical link: https://commits.webkit.org/297883@main
